web/html/src/manager/maintenance/maintenance-windows.js

```diff
@@ -0,0 +1,249 @@
+        this.state.type === "schedule" && this.getCalendarNames();
+        window.addEventListener("popstate", () => {
+            this.updateView(getHashAction(), getHashId());
+        });
+    }
+
+    updateView(action, id) {
+        if (action === "details" && id) {
+            this.getDetails(id, "details");
+        } else if (id || !action) {
+            this.handleForwardAction();
+        } else {
+            this.setState({action: action});
+        }
+        this.clearMessages();
+    }
+
+    getMaintenanceSchedules = () => {
+        const endpoint = "/rhn/manager/api/maintenance/" + this.state.type + "/list";
+        return Network.get(endpoint, "application/json").promise
+            .then(schedules => {
+                this.setState({
+                    action: undefined,
+                    selected: undefined,
+                    schedules: schedules
+                });
+            }).catch(this.handleResponseError);
+    };
+
+    getCalendarNames = () => {
+        const endpoint = "/rhn/manager/api/maintenance/calendar";
+        return Network.get(endpoint, "application/json").promise
+            .then(calendarNames => {
+                this.setState({
+                    calendarNames: Array.from(Array(calendarNames.length).keys()).map(id => ({id: Number(id), text: calendarNames[id]}))
+                });
+            }).catch(this.handleResponseError);
+    };
+
+    getDetails(row, action) {
+        const endpoint = "/rhn/manager/api/maintenance/" + this.state.type + "/" + row + "/details";
+        return Network.get(endpoint, "application/json").promise
+            .then(item => {
+                this.setState({
+                    selected: item,
+                    action: action
+                });
+                history.pushState(null, null, "#/" + action + "/" +
+                    (this.state.type === "schedule" ? item.scheduleId : item.calendarId));
+            }).catch(this.handleResponseError);
+    }
+
+    handleDetailsAction(row) {
+        this.getDetails(row, "details");
+    }
+
+    handleEditAction(row) {
+        this.getDetails(row, "edit");
+    }
+
+    toggleActive(schedule) {
+        Object.assign(schedule, {
+            active: !(schedule.active)
+        });
+        this.updateSchedule(schedule);
+    }
+
+    update(item) {
+        return Network.post(
+            "/rhn/manager/api/maintenance/" + this.state.type + "/save",
+            JSON.stringify(item),
+            "application/json"
+        ).promise.then((_) => {
+            const successMsg = <span>{t(
+                (this.state.type === "schedule" ? "Schedule " : "Calendar ") +
+                "successfully" + (this.state.action === "create" ? " created." : " updated."))}</span>
+            const msgs = this.state.messages.concat(MessagesUtils.info(successMsg));
+
+            while (msgs.length > messagesCounterLimit) {
+                msgs.shift();
+            }
```

Comment:
What about `slicing` the array? https://developer.mozilla.org/it/docs/Web/JavaScript/Reference/Global_Objects/Array/slice